### PR TITLE
Fixes for General Additive Models

### LINKED ISSUE #738: General Additive Models (GAM) for Classification learn the logit, not class probability
### System information

_independent of system_

### Issue

The `GAM Binary Classifier` in `ML.NET` learns to predict the *Logit*, but it is not calibrated to produce a probability. This goes against the standard pattern that we have for classifiers in `ML.NET`.

### Source code / logs

See the `Create` function: https://github.com/dotnet/machinelearning/blob/60ae981223e83b174ecaaf528bd51814a6b0835c/src/Microsoft.ML.FastTree/GamTrainer.cs#L140

### LINKED ISSUE #739: General Additive Model features are not centered after learning.
### Issue

In the standard formulation of GAMs by Rich Caruana's research group [[Lou et al. 2012](http://www.cs.cornell.edu/~yinlou/papers/lou-kdd12.pdf)], the feature responses are centered around the average response in the training set, and an intercept term is added to capture the average response.

In the ML.NET formulation of GAMs, the feature values are not centered.

This is an issue when scoring terms with missing values, as those terms will then deviate from the average by an un-calibrated amount.

### LINKED ISSUE #740: General Additive Models do not allow pruning with a validation set
### Issue

In the standard formulation of GAMs by Rich Caruana's research group [[Lou et al. 2012](http://www.cs.cornell.edu/~yinlou/papers/lou-kdd12.pdf)], the features are pruned post-learning based on a validation set.

In ML.NET, pruning is not allowed.

### LINKED ISSUE #741: FastTree Gradient of Logistic Loss prohibits small learning rates
### Issue

The gradient of the Logistic Loss implemented in `FastTree` uses the `LambdaRank`-style `sigmoid` parameter, set to the learning rate. This quashes the gradients for small learning rates. While this works well for classification tasks, when used by the `General Additive Model` (`GAM`) trainer, it prohibits learning with small learning rates. However, the `GAM` learning-by-boosting technique implemented here requires small learning rates to be stable.

### LINKED ISSUE #742: General Additive Model (GAM) has no summary to extract features at runtime
### Issue

The `Generalized Additive Model` (`GAM`) learner provides a save-as-text option [1], but no option to view the summary at runtime.

[1] https://github.com/dotnet/machinelearning/blob/60ae981223e83b174ecaaf528bd51814a6b0835c/src/Microsoft.ML.FastTree/GamTrainer.cs#L820

### PR DESCRIPTION
**Pull request by rogancarr**
This PR addresses a number of issues with the `General Additive Model` (`GAM`) trainer. In particular, it addresses the issues with the `GAM Classifier` not fitting nor producing a probability, and adds support for validation pruning, summary text, and centering of the feature effects around a mean response (e.g. intercept).

Additionally, the PR addresses some minor issues in the codebase, like `GAMs` using copy-and-paste versions of `FastTree` code, unnecessarily `public` attributes, unused arguments, and splits the `BinaryClassifier` and `Regressor` into separate files.

The changes are as follows:
1. Centered the response and added an intercept term; fixed corresponding issues with table lookups, sparse calculations, and exports. (#739)
2. Added in a validation set and pruning based on the validation set, producing a final pruned graph at the end. (#740)
  a. Switched to using `ScoreTrackers` to keep track of scores during boosted
  b. Save boost iterations as individual graphs (n_features x n_iterations x n_boosts) for pruning.
3. Fixed `GAM Classifier` to use a small learning rate (#741)
  Updated the `FastTreeBinaryClassification` logistic loss gradient to take the `sigmoid` parameter as input: `GAM Binary Classifier` now uses unity; `FastTree Binary Classifier` uses the same default of `2*learning rate` (no change). Optionally, we can plumb this to the FastTree arguments if we show an experimental gain; We can also experiment to see if the `sigmoid` parameter gives gains for GAMs (i.e. by slowing learning even more).
7. Added calibration to the `GamClassifier` to produce probabilities (#738).
  This meant changing the various interfaces.
8. Refactored `GAMRegressor` and `GAMClassifier` into their own files.
  The one file for all GAM trainers had gotten a bit long, and this change is consistent with ML.NET tradition.
9. Added tests to verify train loss and validation metrics (e.g. same value is produced by boosting and the lookup table).
10. Added a Summary to the `GAMPredictor` to produce training statistics and the feature table (#742).
11. Rewrote the core split finding routine to use the same code as FastTree to rely on FastTree unit tests
  The `GAM` routines used a copy-and-paste of internal `FastTree` components. To fight entropy, these were refactored to use the same central calculation, such that it is verified and validated by the `FastTree` unit and end-to-end tests.
12. Removed unused arguments.
13. Allowed for weighted samples.

Fixes #738 
Fixes #739 
Fixes #740 
Fixes #741 
Fixes #742

### CODE REVIEW

File: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj [hunk 1]
```diff
@@ -27,6 +32,8 @@
     <Compile Include="FastTreeRegression.cs" />
     <Compile Include="FastTree.cs" />
     <Compile Include="FastTreeTweedie.cs" />
+    <Compile Include="GamClassification.cs" />
+    <Compile Include="GamRegression.cs" />
     <Compile Include="GamTrainer.cs" />
```
Comment (TomFinley) [inline src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj]: Actually this is somewhat peculiar. I know you didn't do this, but why is the explicit listing in this project necessary at all? Does anything break if you get rid of this entire `ItemGroup` with all the `<Compile Includes`?

**Comment by TomFinley** [comment]
Thanks for doing this @rogancarr -- should I infer from the lack of baseline changes that no tests were migrated to test GAM? I think we probably ought to do some here -- we have some standard regression problems and whatnot.

File: test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        ///     then the training loss is then based on the resulting graph, so the test will
+        ///     not validate that the end graph was constructed correctly.
+        /// </summary>
+        [Fact]
+        [TestCategory("GAM")]
+        public void TestBinaryClassificationGamTrainer()
+        {
+            using (var env = new TlcEnvironment())
+            {
+                var trainFile = "binary_sine_logistic_10k.tsv";
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs]: >binary_sine_logistic_10k [](start = 33, length = 24)

Just a reminder what you need to check in this file into repo, or use build actions to download it.
(if you want to check it in, you also need CELA approval)

Comment (TomFinley) [reply test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs]: Hi @Ivanidzo4ka , I believe that this is a synthetic dataset, so probably CELA is not necessary, since it does not come from any real external source? Or am I wrong?

**Comment by TomFinley** [comment]
> 	<EnableDefaultCompileItems>False</EnableDefaultCompileItems>

I think this is the culprit here, that is requiring the explicit listing.

---
Refers to: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj:7 in d0d4e6b. [](commit_id = d0d4e6b76eacd1873d3d463635c18aa7277603f1, deletion_comment = False)

File: test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs (hunk 2 above)
Comment (rogancarr) [reply test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs]: This is a synthetic dataset. If we want to add this test to the repo, then I'll add the file.

**Comment by rogancarr** [comment]
Yes, my best guess is that this line:
```<EnableDefaultCompileItems>False</EnableDefaultCompileItems>```
is why we are explicitly listing the files to compile.

File: test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs (hunk 2 above)
Comment (rogancarr) [reply test/Microsoft.ML.Predictor.Tests/TestGamTrainer.cs]: But more important question is: Do we need this test?

This is really a test on the internal state of the learner that I used for debugging. If we add baseline tests for `GAMs`, then it seems like we don't need this test, and we may not want it, since it relies on the internal state.

**Comment by rogancarr** [comment]
New commit addressing offline reviews:
- Switched sub-graph calculation from multidimensional arrays to structs
- Removed TrainingInfo (and therefor the test to validate training loss): We need a general way to do this and not a one-off solution for one learner.
- Removed post-training scoring update: At this point it's not necessary and hurts performance
- Made the statistics calculator interface internal to FastTree: It cannot be an abstract class

Next:
- Running performance comparisons against the previous version for GAM{Regressor, Binary Classifier} and FastTree{Regressor, and Binary Classifer}
- Recomputing baselines

File: src/Microsoft.ML.FastTree/Dataset/FeatureFlock.cs [hunk 3]
```diff
@@ -248,7 +248,7 @@ public void CopyFeatureHistogram(int subfeatureIndex, ref PerBinStats[] hist)
             double sumGTTargets = 0.0;
             double sumGTWeights = eps;
             int gtCount = 0;
-            sumWeights = 2 * eps;
+            sumWeights += 2 * eps;
             double gainShift = leafCalculator.GetLeafSplitGain(totalCount, sumTargets, sumWeights);
```
Comment (TomFinley) [inline src/Microsoft.ML.FastTree/Dataset/FeatureFlock.cs]: Nice

**Comment by rogancarr** [comment]
New commits
- Fixed an introduced bug in the split finding algorithm for weighted calculations
- Updated the baseline tests for GamBinaryClassifier

I performed two performance tests, and sent results to reviewers offline:
1. Determine if the new implementation of GAMs had an effect on goodness-of-fit (AUC, RMS) and training time
2. Determine if the changes to FastTree had an effect on goodness-of-fit (AUC, RMS) and training time.

Test Results:
1. GAM changes resulted better goodness of fit for Binary Classification, and equivalent for Regression. Regression was not numerically the same because the order of additions to the final graphs changed. The changes also introduced a slowdown on the order of 1.3 due to the way that graphs are held for pruning (many small pieces instead of one master graph). This can be targeted for future optimization (I have a few ideas).
2. The changes had no measurable effect on FastTree. Goodness-of-fit was numerically identical and training time was equivalent.

**Comment by rogancarr** [comment]
In the build, [SDCA failed a test on Linux Debug](https://dnceng.visualstudio.com/public/_build/results?buildId=20807&view=ms.vss-test-web.test-result-details) -- this seems unrelated to the changes in this PR -- is it possible to rerun the tests?

File: src/Microsoft.ML.FastTree/GamRegression.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+namespace Microsoft.ML.Runtime.FastTree
+{
+    public sealed class RegressionGamTrainer :
+    GamTrainerBase<RegressionGamTrainer.Arguments, RegressionGamPredictor>
+    {
+        public partial class Arguments : ArgumentsBase
+        {
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Metric for pruning. (For regression, 1: L1, 2:L2; default L2)", ShortName = "pmetric")]
+            [TGUI(Description = "Metric for pruning. (For regression, 1: L1, 2:L2; default L2")]
+            public int PruningMetrics = 2;
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/GamRegression.cs]: >int PruningMetrics = 2 [](start = 19, length = 22)

wow, really? Can we maybe have enum here?

File: src/Microsoft.ML.FastTree/GamTrainer.cs [hunk 5]
```diff
@@ -209,25 +96,41 @@ public abstract class ArgumentsBase : LearnerInputBaseWithWeight
...
         private SufficientStatsBase[] _histogram;
+        private ILeafSplitStatisticsCalculator _leafSplitHelper;
+        private ObjectiveFunctionBase _objectiveFunction;
+        private bool HasWeights => TrainSet?.SampleWeights != null;
 
-        protected bool HasWeights => TrainSet?.SampleWeights != null;
+        // Training Datastructures
+        private SubGraph _subGraph;
 
         //Results of Training
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/GamTrainer.cs]: >//Results of Training [](start = 8, length = 21)

please use the common syntax for comments.
//<space>Sentence starting from capital letter (no other capitalized words inside)<period>

File: src/Microsoft.ML.FastTree/GamTrainer.cs [hunk 6]
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
...
                     ch.Done();
                 }
-            }
 
             InitializeThreads(numThreads);
         }
 
-        public sealed override TPredictor Train(TrainContext context)
+        /// <summary>
+        /// Load the data off the context and train the model
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/GamTrainer.cs]: >Load the data off the context and train the model [](start = 12, length = 49)

That is not a helpful comment :)

File: src/Microsoft.ML.FastTree/GamTrainer.cs [hunk 7]
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
...
+        /// </summary>
+        /// <param name="context"></param>
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/GamTrainer.cs]: >param [](start = 13, length = 5)

remove empty param tags

File: src/Microsoft.ML.FastTree/GamTrainer.cs [hunk 8]
```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
...
-                var pred = CreatePredictor();
+
                 ch.Done();
-                return pred;
             }
         }
 
-        private protected abstract TPredictor CreatePredictor();
+        /// <summary>
+        /// Define the ScoreTrackers to use in the calculation.
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/GamTrainer.cs]: >Define the ScoreTrackers [](start = 12, length = 24)

these comments are not helpful. I already have a good suspicion that `DefineScoreTrackers` defines score trackers. 

These are private methods too, so might as well not have comments at all

Comment (rogancarr) [reply src/Microsoft.ML.FastTree/GamTrainer.cs]: Thanks -- removed!

File: src/Microsoft.ML.FastTree/GamTrainer.cs (hunk 7 above)
Comment (rogancarr) [reply src/Microsoft.ML.FastTree/GamTrainer.cs]: Thanks, -- removed!

File: src/Microsoft.ML.FastTree/GamTrainer.cs (hunk 6 above)
Comment (rogancarr) [reply src/Microsoft.ML.FastTree/GamTrainer.cs]: Removed!

File: src/Microsoft.ML.FastTree/GamTrainer.cs (hunk 5 above)
Comment (rogancarr) [reply src/Microsoft.ML.FastTree/GamTrainer.cs]: Done for all comments!

File: src/Microsoft.ML.FastTree/GamRegression.cs (hunk 4 above)
Comment (rogancarr) [reply src/Microsoft.ML.FastTree/GamRegression.cs]: This is a holdover from `FastTree` -- this is copied over from `FastTreeRegression`.  It's a parameter to the [`RegressionTest` class](https://github.com/dotnet/machinelearning/blob/a8cd341f8c03985aaea216e8c911c466ad020b1e/src/Microsoft.ML.FastTree/Training/Test.cs#L516).

If possible, I'd like to file this as a separate issue and fix as a separate PR.

**Comment by rogancarr** [comment]
Thanks for the review @TomFinley  and @Zruty0 ! I addressed all but one issue and responded to one in the comments here. Note that WIP is flagged, but this is not WIP -- I had a WIP commit that I didn't remove because we will squash-merge this branch.
